src/puppetlabs/puppetdb/http/command.clj

```diff
@@ -77,40 +92,145 @@
+                " (loading into RAM)"))
+          (if-let [body-str (request/body-string req)]
+            (let [body (json/parse-string body-str)                    
+                  required-fields #{"command" "version" "payload"}]
+              (if-not (= required-fields (set (keys body)))
+                (http/error-response
+                 (str "Unexpected POST command fields: "
+                      (str/join " " (set/difference (set (keys body))
+                                                    required-fields))))
+                (let [certname (get-in body ["payload" "certname"])
+                      params (merge params
+                                    (select-keys body ["command" "version"])
+                                    (when certname {"certname" certname}))
+                      {:strs [command version certname]} params]
+                  ;; Enforce certname too? (currently breaks tests)
+                  (if-not (and command version)
+                    (missing-items "elements" command version true)
+                    (handle
+                     (assoc req
+                            :body (json/generate-string (body "payload"))
+                            :params (assoc params "version"
+                                           (Integer. version))))))))
+            (http/error-response "Empty application/json POST body")))))))
+
+(defn- realize-body [body reject-large-commands? max-command-size]
+  "Returns the body as an in-memory string or byte-array, reading it
+  if necessary.  Throws ::body-stream-overflow if the max-command-size
+  is not respected."
+  (if-not reject-large-commands?
+    (if (instance? java.io.InputStream body)
+      (IOUtils/toByteArray body)
+      body)
+    (cond
+      (instance? java.io.InputStream body)
+      (IOUtils/toByteArray
+       (restrained-drained-stream body (long max-command-size)))
+
+      (string? body) ;; Assuming UCS-2
+      (if (> (* 2 (count body)) max-command-size)
+        (throw+ ::body-stream-overflow)
+        body)
+
+      :else
+      (throw (Exception. (str "Unexpected body type: " (class body)))))))
 
 (defn- enqueue-command-handler
   "Enqueues the command in request and returns a UUID"
-  [enqueue-fn get-response-pub]
-  (fn [{:keys [body-string params] :as request}]
-    (let [uuid (kitchensink/uuid)
-          completion-timeout-ms (some-> params
-                                        (get "secondsToWaitForCompletion")
-                                        Double/parseDouble
-                                        (* 1000))
-          do-submit #(enqueue-fn body-string uuid)]
-      (if (some-> completion-timeout-ms pos?)
-        (blocking-submit-command do-submit (get-response-pub) uuid completion-timeout-ms)
-        (do
-          (do-submit)
-          (http/json-response {:uuid uuid}))))))
+  [enqueue-fn get-response-pub reject-large-commands? max-command-size]
```

Comment:
can we imply `reject-large-commands?` with a `nil` `max-command-size`? Not sure if that's good or bad, but at least it's one fewer thing to flow around.